### PR TITLE
Fix serialization of BatchNormalization layer

### PR DESCRIPTION
**Pull request by matthiasplappert**
The serialization of the `BatchNormalization` layer is currently broken. This can be easily tested with the following code:

```
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.layers.normalization import BatchNormalization
from keras.models import model_from_config


def build_model():
	model = Sequential()
	
	model.add(Dense(20, input_shape=(10,)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	
	model.add(Dense(10))
	model.add(Activation('linear'))
	
	model.compile(loss='mse', optimizer='sgd')
	return model


def clone_model(model):
	config = model.get_config()
	clone = model_from_config(config)
	clone.compile(loss='mse', optimizer='sgd')
	clone.set_weights(model.get_weights())
	return clone


model = build_model()
clone = clone_model(model)
```

The same problem would occur if a network with `BatchNormalization` would be written to disk and re-loaded later.

The root of the problem lies in the way the hacky way the layer handles `running_mean` and `running_std`: These are *not* part of the params array since they are not optimized during training. However, they are still part of the weights list obtained by `get_weights()` (see removed implementation). This causes problems since the layer reportedly only has two params (`gamma` and `beta`) but the weights list contains for elements. When de-serializing, this triggers an exception. Additionally, since the layer reports that it only uses two weight tensors, the serialization process only ever passes a list of two elements to `set_weights()`. This is very much problematic since the following layers would then (had the assertion not been there) receive the `running_mean` and `running_std` as weight vectors, corrupting the entire de-serialization process.

My solution moves the `running_mean` and `running_std` from the weights to the configuration. I do realize that this is not optimal since this means that we encode a potentially high-dimensional vector into the configuration. However, it was the easiest solution to fix the problem now.

Any feedback is highly welcome, maybe someone has a better idea than somewhat misusing the config like that.

### CODE REVIEW

**Comment by matthiasplappert** [comment]
The tests need fixing, but I'd like to get some input before on the general approach before spending more time on it.

**Comment by fchollet** [comment]
Sounds reasonable.

**Comment by tzachar** [comment]
I believe a simpler solution is more desirable.

The problem lies with containers.Sequential, which does not have a symmetric implementation of get/set weights: look at https://github.com/fchollet/keras/blob/master/keras/layers/containers.py#L158
The fix is to use len(self.layers[i].get_weights()) instead of len(self.layers[i].params).

Anyway, this will not be triggered in serialization unless a Sequential model is embedded inside a different model, since modles.Sequential handles the load/save weight appropriately.

Cheers.

**Comment by fchollet** [comment]
The reason why the fix proposed here is reasonable is because:

> I do realize that this is not optimal since this means that we encode a potentially high-dimensional vector into the configuration.

In general it will be rather small. For a 128x128 FC layer you get two arrays of 128 floats. For a Conv2D layer with 32 filters you get 2 arrays of 32 floats. Encoding this in JSON with the config would really not be a problem.

**Comment by tzachar** [comment]
And what about a layer of size 4096? And deep architectures of several of
these layers? This does not scale.

The solution I proposed is simpler.

**Comment by matthiasplappert** [comment]
I agree with @tzachar, my current solution is far from optimal. The `running_mean` and `running_std` really should be part of the weights list and therefore be stored in the HDF5 file.

I have an alternative proposal to solve this problem, which would also extend to the needs of other layers that might run into a similar problem in the future. Instead of storing the `running_mean` and `running_std` tensors in the `params` list, we add an additional list (e.g. `additional_params` or something more descriptive). These additional params are then serialized and de-serialized in the `set_weights`/`get_weights` methods in the core layer class. However, they are not used in the optimizer since they are not part of the `params` list.

In summary, we would have different lists for parameters that are used by the optimizer and therefore need a gradient (e.g. the standard weight matrix of a layer) and an additional list of parameters that are updated through other means (e.g. the `running_mean` and `running_std` parameters). The `set_weights`/`get_weights` methods would then operate on the concatenated list of `params` and `additional_params`.

These changes should be very easy to make and, I think, are a far better solution. I'm happy to do the necessary work.

**Comment by matthiasplappert** [comment]
I've implemented the alternative fix and I like it a lot better: https://github.com/matthiasplappert/keras/commit/49b06628ae3e0965315fb81d59cf463715b06c22 What do you guys think? I'll squash the commits as soon as we've decided on one approach.

**Comment by fchollet** [comment]
I think it's a reasonable approach. Conceptually it adds quite a bit of complexity, though. I'm wondering if we could unify this with layer `updates` in some way.

Not a fan of the name `additional_params`, which contributes to the "hacky special case" impression...

**Comment by matthiasplappert** [comment]
I agree, `additional_params` is not a good name. I could't come up with something better though.

I don't know enough about the inner workings of keras to understand your suggestion regarding `update`. Could you elaborate?

**Comment by fchollet** [comment]
> I don't know enough about the inner workings of keras to understand your suggestion regarding update. Could you elaborate?

We are currently storing parameter updates that are independent from the global optimization step in a layer attribute `updates`. 

I still don't see the "right" way to do this. The previous solution might have be better.

> This does not scale.
 
It scales linearly, and even for a state of the art inception 7 architecture with batch norm after every convolution the resulting JSON / YAML would be perfectly manageable.

Conceptually, it's preferable to store weights in hdf5, though. But there is not big technical obstacle here. 

I'll think about it a bit more.

**Comment by AvantiShri** [comment]
Randomly chiming in as a user: we occasionally find ourselves manually inspecting the json/yaml to figure out the exact model specs something was run with. If there were a long string of weights in their, it might make the files a lot more tedious to inspect.

**Comment by fchollet** [comment]
The correct solution will require refactoring the internal API, which I will do asap. For now, we will go with your fix, but please rename `additional_params` to `non_trainable_weights`.

Thanks!

**Comment by matthiasplappert** [comment]
I've renamed the attribute and rebased against master. All tests are passing locally, so I expect them to also pass on CI. Should be good to go.

File: keras/layers/core.py [hunk 1]
```diff
@@ -33,6 +33,11 @@ class Layer(object):
             (e.g. `(32, 100)` for a batch of 32 100-dimensional inputs).
     '''
     def __init__(self, **kwargs):
+        if not hasattr(self, 'params'):
```
Comment (matthiasplappert) [inline keras/layers/core.py]: This change is necessary because we need to ensure that `non_trainable_weights` exists before applying the `kwargs`. Otherwise, some layers internally call `set_weights`, which in turn fails because `non_trainable_weights` doesn't exist. It only seemed natural to also move the check for `params` up to the same place.

**Comment by fchollet** [comment]
This has been merged: https://github.com/fchollet/keras/commit/ff4a9b7b24011e5eec09bbf03c7b3afe8f1a958c
